### PR TITLE
DEV-127 / DEV-126: Pin the headers with CSS while the window scrolls the grid

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR fixes the headers flickering while the window scrolls a grid: the row headers sideways (DEV-127) and the column headers up and down (DEV-126).

A grid with no `width` and no `height` is scrolled by the window. In that mode the browser scrolls the page on its compositor, and the `scroll` listener moved the clones one frame later – a `translate3d` on most of them, a `left` or `bottom` inset on the bottom ones. On about every other frame of a wheel scroll the headers were torn away from the viewport edge or missing altogether: 61 of 121 painted frames for the row headers on Chromium, and 49 of 98 for the column headers. Every engine is affected. A grid that scrolls inside its own box is not: its wheel listener scrolls the holder from JavaScript, so the clones move in the same frame.

The fix pins every clone the window carries with `position: sticky`, which the browser resolves on the scroll's own frame. The two axes are one mechanism: a rail spans the table on each axis its clone travels, and each overlay names the axes it follows – the inline-start one sideways, the top and bottom ones up and down, the corners both.

- A sticky box only travels inside its parent, and every ancestor of the clones is as large as the viewport, not the table. While the window owns an axis and the overlay renders, the clone sits in a `div.htOverlayRail` (`walkontable/src/overlay/overlayRail.ts`) that is as wide as the table when the clone travels sideways and as tall as the table when it travels up and down, and it is `position: sticky` inside it; otherwise it stays where it was. A rail that spans the block axis lies over the cells, so it takes no pointer events and the clone takes its own back. Whatever the rail does not hold stays the listener's transform, which is what a corner uses when an element owns its other axis.
- A sticky box only shifts from where it would otherwise stand, so a clone anchored to the bottom has to stand at its rail's bottom: as the rail's only child it stands at the top, where a `bottom: 0` inset never engages and the frozen rows sit under the column headers. The rail is a column flex box for that case. Its height also reaches the table's painted bottom rather than its CSS-integer one, which is the fractional-zoom correction the bottom overlay used to apply to its own inset.
- Moving a clone into or out of its rail detaches it for a moment. The focus manager focuses the topmost copy of a cell, which for a frozen-column or row-header cell is inside that clone, so the rail records the focused element and gives the focus back, as row recycling does in `render/rows.ts`.
- A sticky shift is part of the layout, where a transform never was, so `offsetLeft`, `offsetTop` and the `offset()` helper now see it. `BaseEditor#getEditedCellRect` therefore adds `Overlay#getOverlayTransformOffset()` (0 while a rail holds that overlay's own axis) instead of the full overlay offset, on both axes; with the full offset the editor of a frozen-column or frozen-row cell opens a scroll distance away. `getOverlayOffset()` keeps its meaning.
- Code that recognizes a clone by its parent looks through the rail. `isInternalElement()` steps over one when it checks that a clone belongs to the grid, which copy/paste, the text editor and the focus manager ask; `Border#getDimensionsFromHeader` reads the grid root's selection classes through the master instead of its own table's parent. The rail's class name lives in `overlay/constants.ts`, so the DOM helper and the rail do not import each other.
- The row-header seam rule in `_base.scss` reaches clones through `.ht_master ~`; it also reaches them inside a rail.
- An overlay that stops rendering leaves its rail. `Overlay#reset()` releases it, so a clone whose width is cleared shrinks to 0 as before instead of stretching across the rail, and the corners – which are repositioned on every draw, rendering or not – pin only while they render, so the next draw cannot put an idle corner back.

At rest and after a scroll, the grid renders pixel-identical: 36 of 36 compared screenshots (3 themes × LTR/RTL × with/without frozen panes × at rest / after a sideways page scroll / after a vertical one) differ by 0 pixels, against `develop` for the row-header change and against that build for the column-header one. One layout changes by design: on a page wider than the grid, scrolled past the grid's inline-end edge, the headers stay against the table's end instead of jumping back to its start. In that regime `getOverlayOffset()` still reports 0, which legacy specs assert, so anything placed over the clone from that number – the `manualColumnMove` backlight over a frozen column – is off by the parked distance there.

`walkontable/AGENTS.md` documents the mechanism and its rules.

### Test evidence (required for source changes)

- Unit tests added/modified (`*.unit.js`): `handsontable/src/3rdparty/walkontable/test/unit/overlay/overlayRail.unit.ts` (new, 19 tests, covering both axes, the focus restore and the click-through rail); `handsontable/src/helpers/dom/__tests__/element.unit.ts` (4 new `isInternalElement` tests).
- E2E tests added/modified (Playwright `tests/e2e/*.spec.ts`): `tests/e2e/walkontable/window-scroll-pinned-overlays.spec.ts` (new, 21 tests), with `tests/fixtures/pages/walkontable/WindowScrollPinnedOverlaysPage.ts` and `tests/fixtures/demo/walkontable/window-scroll-pinned-overlays.html`. The pinning tests record a real wheel scroll through a CDP screencast and scan every painted frame for the color the fixture paints each clone – the DOM reports the clone at the edge even on a torn frame, so no DOM assertion can see this bug. A positive control on each axis, with the clones deliberately unpinned, proves the scan sees a clone left behind. The fixture is wide and short for the sideways legs and tall and narrow for the up-and-down ones, so one wheel moves one axis. The placement tests cover the editor, the fill handle and the row-resize handle over frozen columns after a horizontal page scroll, LTR and RTL, and the editor over a frozen bottom row after a vertical one.
- Type tests (`*.types.ts`) updated if public API changed: none – no public API change.
- For a bug fix – the spec that fails without this fix: `the row headers and the corner stay pinned on every painted frame` and its frozen-panes and RTL variants (on `develop`: 49 of 98 painted frames torn for each of the three clones), and `the column headers stay pinned on every painted frame` with `frozen rows and both corners stay pinned on every painted frame` (49 of 98 with the listener positioning `develop` uses). With the editor change alone reverted, `the editor opens over a frozen-column cell` fails in LTR and RTL, and `the editor opens over a frozen-top-row cell after a vertical page scroll` lands 3200px below the cell while every other editor test stays green. `a corner that stops rendering leaves its rail and stays out of it` fails without the corners' render gate. The two focus unit tests fail without the restore, and `should step over the rail and no further` fails on a rail step that climbs one level too far. The legacy specs `rowHeader.spec.js`, `settings/fixedColumnsStart.spec.js` and `helpers/dom/__tests__/element.spec.js` fail without the `reset()` release and the `isInternalElement` rail step.
- Demo page / recorded trace (for UI changes): none.
- Not verified: real Safari (WebKit was measured through Playwright), and trackpad momentum scrolling (the tests drive discrete wheel steps). Selection borders in the pinned clones have no test of their own: header and cell extents coincide except in setups such as nested headers, so the `Border` root lookup is a correctness fix without an observable difference to assert on this fixture.

Frames with the row headers torn, per scrolling frame of the same horizontal wheel scroll:

| Engine | `develop` | This PR |
|---|---|---|
| Chromium 151 (lossless CDP screencast) | 61 / 121 | 0 / 109 |
| WebKit 26.5 (recorded video) | 8 / 53 | 0 / 58 |
| Firefox 153 (recorded video) | 30 / 61 | 4 / 55 |

And the same for the column headers, on a vertical wheel scroll of the spec's own fixture (Chromium, lossless CDP screencast): 49 of 98 painted frames with the listener positioning, 0 of 98 with this change.

The 4 remaining Firefox frames blank the whole inline-start strip, cells included – the rendered column band trailing a fast scroll – not the row headers.

### Commands run

```bash
npm --prefix handsontable run build
cd tests && HOT_TEST_PORT=8137 npx playwright test --project=e2e-main
```
1587 passed, 1 skipped

```bash
npm --prefix handsontable run test:unit
```
Test Suites: 405 passed, 405 total – Tests: 5590 passed, 5590 total

```bash
npm --prefix handsontable run test:walkontable
```
851 specs, 0 failures

```bash
npm --prefix handsontable run test:e2e -- "--testPathPattern=(helpers/dom/__tests__/element\.spec|plugins/copyPaste|editors/textEditor|editors/baseEditor|focusManager|focusHandling|__tests__/rowHeader\.spec|__tests__/tableView\.spec|settings/fixedColumnsStart|settings/fixedRowsTop|settings/fixedRowsBottom|settings/preventOverflow|settings/width|settings/height|settings/colHeaders|settings/rowHeaders|rowHeaderSelection|plugins/manualColumnFreeze|plugins/manualRowResize|plugins/manualRowMove|plugins/manualColumnMove|plugins/nestedHeaders|plugins/collapsibleColumns|test/e2e/Core_view|[Ss]election|customBorders|scroll)"
```
3001 specs, 0 failures

The rest of the legacy suite ran green on the row-header change against an earlier `develop` base; since then only the set above has been re-run, and CI runs the whole suite.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):

1. DEV-127
2. DEV-126

### Affected project(s):

- [x] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://cla.handsontable.com/sign) — one signature covers both Handsontable and HyperFormula; the `cla/signed` check on this PR confirms it.
- [ ] My change requires a change to the documentation.
- [ ] MANUAL QA NEEDED — 

ClickUp tasks: https://app.clickup.com/t/9015210959/DEV-127 and https://app.clickup.com/t/9015210959/DEV-126

---

> [!NOTE]
> **Medium Risk**
> Touches Walkontable overlay DOM structure, positioning math, and focus/copy-paste paths; behavior is heavily tested but limited edge cases remain (e.g. scroll past table end vs manual-move backlight).
> 
> **Overview**
> Fixes **frozen header flicker** when the **window** scrolls an unsized grid (DEV-127 / DEV-126) by stopping scroll-listener `transform`/`inset` updates that paint one frame behind the compositor.
> 
> **Window-scroll mode** now pins overlay clones with **`position: sticky`** inside new **`div.htOverlayRail`** wrappers (`overlayRail.ts`), wired from top/bottom/inline-start overlays and corners via `Overlay#getRail()` / `pin()` / `release()`. Element-scrolled grids still use the old positioning; rails release on `reset()` and when overlays stop rendering.
> 
> Because sticky shifts are in layout, callers that used to add `getOverlayOffset()` on top of `offset()` now use **`getOverlayTransformOffset()`** (e.g. `BaseEditor#getEditedCellRect`). **`isInternalElement`** steps over rails so frozen cells stay “internal”; selection **`Border`** reads the master root; **`_base.scss`** reaches clones inside rails.
> 
> Adds **unit tests** for `OverlayRail` / `isInternalElement` and **Playwright E2E** with CDP screencast frame scans plus editor/fill-handle/resize placement after page scroll (LTR/RTL).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2f411f57cea3438bf3a925a6bf226280146c299b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [DEV-127 The rowHeader flickers while scrolling horizontal on Firefox and Safari](https://app.clickup.com/t/9015210959/DEV-127)

**Comment by qunabu** [comment]
Task linked: [DEV-126 The colHeader flickers while scrolling vertical](https://app.clickup.com/t/9015210959/DEV-126)

File: handsontable/src/editors/baseEditor/baseEditor.ts [hunk 1]
```diff
@@ -709,12 +709,16 @@ export class BaseEditor {
     // When the scrollable element is Window object then the editor position needs to be compensated
     // by the overlays' position (position relative to the table viewport). In other cases, the overlay's
     // position always returns 0.
+    // Only the part the offset chain cannot see: a clone pinned with `position: sticky` is already
+    // shifted in the layout `offset()` walks (DEV-126 on this axis).
     if (['top', 'top_inline_start_corner'].includes(overlayName)) {
-      topPos += wtOverlays.topOverlay?.getOverlayOffset() ?? 0;
+      topPos += wtOverlays.topOverlay?.getOverlayTransformOffset() ?? 0;
```
Comment (Haxikowy) [inline handsontable/src/editors/baseEditor/baseEditor.ts:715]: Only `top`/`top_inline_start_corner`/`inline_start` get `getOverlayTransformOffset()` here; `bottom` and `bottom_inline_start_corner` never do, even though `bottomOverlay.ts` now rails on the same block axis. Editing a cell in a frozen bottom row (or that corner) while the window scrolls will open the editor a scroll-distance away from the cell.

File: tests/e2e/walkontable/window-scroll-pinned-overlays.spec.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    // non-passive, so the page is scrolled from JavaScript here and no frame could tear either way.
+    clone = await wt.box(wt.inlineStartClone);
+    expect(Math.abs(clone.x)).toBeLessThanOrEqual(PLACEMENT_TOLERANCE);
+
+    const corner = await wt.box(wt.topCornerClone);
+
+    expect(Math.abs(corner.x)).toBeLessThanOrEqual(PLACEMENT_TOLERANCE);
+  });
+
+  test.describe('scrolling the page up and down', () => {
```
Comment (Haxikowy) [inline tests/e2e/walkontable/window-scroll-pinned-overlays.spec.ts:171]: This vertical-scroll block only tests the editor; the fill-handle and row-resize-handle tests exist only in the horizontal block above (lines 119, 132). The PR description implies both handles are covered on frozen rows too, but that axis is untested.

File: handsontable/src/3rdparty/walkontable/src/overlay/overlayRail.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+  pinsBlock(): boolean {
+    return this.#pinsBlock;
+  }
+
+  /**
+   * Puts the clone in its rail, or updates the rail of a clone already in one.
+   *
+   * @param {RailPlacement} placement What the rail spans and which axes the clone travels.
+   */
+  pin(placement: RailPlacement): void {
```
Comment (Haxikowy) [inline handsontable/src/3rdparty/walkontable/src/overlay/overlayRail.ts:151]: `pin()` rewrites ~10-15 style properties on the rail and clone every time it's called, even when the placement hasn't changed since the last draw. Since `resetFixedPosition()` calls this unconditionally on every full draw (not just on scroll), this is a real per-render cost increase over the old 1-2-property transform/inset write.

File: handsontable/src/3rdparty/walkontable/src/overlay/regions/_base.ts [hunk 4]
```diff
@@ -375,6 +381,55 @@ export abstract class Overlay {
...
+   * The part of {@link Overlay#getOverlayOffset} that the layout does not already carry.
+   *
+   * A reader that places something from a clone element's document position (`offsetLeft`, the
+   * `offset()` helper) has to add the overlay offset only when the clone is moved by something that
+   * position cannot see – a transform. A clone pinned by its rail is shifted by `position: sticky`,
+   * which IS in the layout, so adding the offset again would count the scroll twice.
+   *
+   * @returns {number}
+   */
+  getOverlayTransformOffset(): number {
```
Comment (Haxikowy) [inline handsontable/src/3rdparty/walkontable/src/overlay/regions/_base.ts:428]: For a corner overlay (`railAxis` stays `null` from the base class), this ternary's condition is false and falls into the `rail?.pinsBlock()` branch instead of a neutral answer, so a block-pinned corner would silently get `0` instead of its real offset. Dead code today since only `topOverlay`/`inlineStartOverlay` call this, but worth making the `null` case return `false` explicitly before a corner is wired to it.

File: handsontable/src/3rdparty/walkontable/src/selection/border/border.ts [hunk 5]
```diff
@@ -1866,7 +1866,9 @@ class Border {
     containerOffset: { top: number; left: number }): false | [HTMLElement, number, number] {
     const { geometryReader } = this.wot.domBindings;
     const { wtTable } = this.wot;
-    const rootHotElement = wtTable.wtRootElement.parentNode as HTMLElement;
+    // Read off the master: a pinned clone's parent is its rail, not the grid root (DEV-127).
+    const masterTable = this.wot.cloneSource?.wtTable ?? wtTable;
```
Comment (Haxikowy) [inline handsontable/src/3rdparty/walkontable/src/selection/border/border.ts:1870]: This fix runs unconditionally for every clone's `Border` now and has no test of its own per the PR body. `nestedHeaders` + frozen panes + window-scroll + a header-anchored row/column selection is an untested combination that exercises exactly this code path (`getHeaderFn`/`dimensionFn` walk header rows by level).

**Review by Haxikowy — COMMENTED** [review]
## Findings with no inline home

**High — sticky silently fails on transformed/filtered ancestors.** `overlay/regions/_base.ts`'s axis-owner check (`#computeMainScrollableElement`) only inspects `overflow`/`overflow-x`/`overflow-y`. It never looks at `transform`, `filter`, `contain`, or `will-change: transform`/`perspective` on an ancestor, and any of those defeat `position: sticky` with no warning. A GPU-accelerated card, a CSS-animated modal, or a `will-change` hint around the grid brings back the torn-header bug this PR fixes, just silently instead of every other frame. Nothing here — code, tests, or the new AGENTS.md section — accounts for it.

**High — undisclosed `manualRowMove` regression.** `manualRowMove.ts:451` reads the raw `getOverlayOffset()`, same as `manualColumnMove.ts:683` — the pattern the PR's own AGENTS.md addition calls "off by the parked distance" once scrolled past the table's end. The PR body names only `manualColumnMove`. `manualRowMove`'s vertical backlight over a frozen row has the same regression, untested and unmentioned.

**Medium — new DOM wrapper breaks selector-based customer CSS, undocumented.** `_base.scss` needed a parallel `.ht_master ~ .htOverlayRail > .handsontable` selector to keep the row-header seam rule working once clones move into a rail. Customer CSS built on `.ht_master ~ .handsontable` sibling selectors targeting overlay clones directly breaks the same way, silently. The PR's documentation checkbox is unchecked.

**Medium — `touchScroll` caches clone-parent references the rail now reparents.** `touchScroll.ts:148-161` caches `overlay.clone?.wtTable.holder.parentNode` and toggles CSS-transition classes on those nodes during touch scroll. `OverlayRail#pin()`/`release()` reparent those same nodes on draws where pin state changes, which can cut a transition short mid-scroll. `touchScroll` has no tests of its own, so check this on mobile Safari before shipping.

**Medium — manual QA gap left unacknowledged.** The fix targets a compositor-timing bug described as affecting every engine, but real Safari and trackpad momentum scrolling are admitted as untested. The MANUAL QA NEEDED checkbox is unchecked, no `Requires Manual QA` label, no note on why the CDP-screencast evidence covers this.

**Low — `Overlay#destroy()` doesn't release its rail.** `reset()` calls `#rail?.release()`; `destroy()` doesn't. Harmless today because `Core#destroy` wipes the whole `rootContainer` afterward regardless, but a latent trap if a future teardown path ever removes only the overlay clones.

**Low — possible forced-layout ordering risk, unmeasured.** `topOverlay`/`inlineStartOverlay` moved from transform (paint-only) to sticky (layout-affecting) writes; `bottomOverlay`'s `getBoundingClientRect()` read runs right after, in the same draw. If the writes and reads aren't batched, that's a forced-reflow site. Not profiled — a candidate for whoever traces it.

File: handsontable/src/editors/baseEditor/baseEditor.ts (hunk 1 above)
Comment (budnix) [reply handsontable/src/editors/baseEditor/baseEditor.ts:715]: Fair point, but this one is on purpose: the compensation exists only for a clone the offset chain cannot see, and the bottom one has always been placed by an inset (`bottomOverlay.ts` on develop), now a sticky shift — both are in `offsetTop`. `the editor opens over a frozen-bottom-row cell after a vertical page scroll` covers it, and I added the scroll-distance and in-rail preconditions in 2f411f57ce so it cannot pass vacuously.

File: tests/e2e/walkontable/window-scroll-pinned-overlays.spec.ts (hunk 2 above)
Comment (budnix) [reply tests/e2e/walkontable/window-scroll-pinned-overlays.spec.ts:171]: Good point about the axis gap. I wrote the vertical fill-handle test and then dropped it — it compares the handle with its own cell in the same clone, so a shared offset cancels: shifting the anchor 40px left all three fill-handle tests green, the existing horizontal ones included. Strengthened the two vertical editor tests in 2f411f57ce instead, since those do cross clones.

File: handsontable/src/3rdparty/walkontable/src/overlay/overlayRail.ts (hunk 3 above)
Comment (budnix) [reply handsontable/src/3rdparty/walkontable/src/overlay/overlayRail.ts:151]: Fair point. It only runs on window-scrolled grids, and only for the clones the window actually carries — but I have not profiled the per-draw cost, so I would rather measure it than guess. Leaving it for now.

File: handsontable/src/3rdparty/walkontable/src/overlay/regions/_base.ts (hunk 4 above)
Comment (budnix) [reply handsontable/src/3rdparty/walkontable/src/overlay/regions/_base.ts:428]: Nice catch — fixed in 2f411f57ce. `railCarriesAxis()` now answers `false` when the overlay names no axis, with a unit test for the corner case.

File: handsontable/src/3rdparty/walkontable/src/selection/border/border.ts (hunk 5 above)
Comment (budnix) [reply handsontable/src/3rdparty/walkontable/src/selection/border/border.ts:1870]: Good spot, that is the combination that would show it. Not added yet — after the fill-handle test above turned out unable to fail, I would rather land this one with a proven negative control than add another test that cannot.

**Comment by Haxikowy** [comment]
The inline threads are settled now. These 7 were only in the review body, not attached to a diff line, and none have a response yet:

1. `overlay/regions/_base.ts`'s axis-owner check only inspects `overflow`/`overflow-x`/`overflow-y`. `transform`, `filter`, `contain`, and `will-change: transform`/`perspective` on an ancestor defeat `position: sticky` with no warning, and none of those show up in that check. A GPU-accelerated card, a CSS-animated modal, or a `will-change` hint around the grid brings back the torn-header bug, just silently instead of every other frame.

2. `manualRowMove.ts:451` reads the raw `getOverlayOffset()`, same as `manualColumnMove.ts:683` — the pattern the AGENTS.md addition calls "off by the parked distance" once scrolled past the table's end. The PR body names only `manualColumnMove`; `manualRowMove`'s vertical backlight over a frozen row has the same regression, untested and unmentioned.

3. `_base.scss` needed a parallel `.ht_master ~ .htOverlayRail > .handsontable` selector to keep the row-header seam rule working once clones move into a rail. Customer CSS built on `.ht_master ~ .handsontable` sibling selectors breaks the same way, silently, with nothing documented.

4. `touchScroll.ts:148-161` caches `overlay.clone?.wtTable.holder.parentNode` and toggles CSS-transition classes on those nodes during touch scroll. `OverlayRail#pin()`/`release()` reparent those same nodes on pin-state-change draws, which can cut a transition short mid-scroll. `touchScroll` has no tests of its own to catch it.

5. Real Safari and trackpad momentum scrolling are admitted as untested, for a compositor-timing bug described as affecting every engine. MANUAL QA NEEDED is unchecked and there's no `Requires Manual QA` label.

6. `reset()` calls `#rail?.release()`; `Overlay#destroy()` doesn't. Harmless today since `Core#destroy` wipes `rootContainer` afterward regardless, but a latent trap if a future teardown path ever removes only the overlay clones.

7. `topOverlay`/`inlineStartOverlay` moved from transform (paint-only) to sticky (layout-affecting) writes; `bottomOverlay`'s `getBoundingClientRect()` read runs right after, in the same draw. Not profiled — a candidate for whoever traces it, not a confirmed issue.

Take a pass on these, or say which ones you're punting on.

**Comment by budnix** [comment]
@Haxikowy Thanks — taking all seven. Three I measured, two I am pushing back on, one I tried and reverted, one is the author's call.

**1. Sticky under transformed/filtered ancestors — measured, does not reproduce.**

This is `position: fixed` behavior. A transform makes an ancestor the containing block for fixed and absolute descendants; a sticky box is constrained by its nearest *scrolling* ancestor, which is what the axis-owner check already resolves.

I wrapped the grid in an ancestor carrying each property in turn and scrolled the window 600px, reading the top clone's viewport position:

| ancestor | clone top at scroll 0 | at scroll 600 |
|---|---|---|
| none (control) | 0 | 0 |
| `transform: translateZ(0)` | 0 | 0 |
| `transform: translateX(0)` | 0 | 0 |
| `will-change: transform` | 0 | 0 |
| `filter: blur(0px)` | 0 | 0 |
| `contain: paint` | 0 | 0 |
| `perspective: 1000px` | 0 | 0 |
| **clones unpinned (negative control)** | 0 | **−600** |

The negative control moves, so the probe can see a failure. None of the six do.

**2. `manualRowMove` — the code is right; the PR description is what is short.**

Swapping in `getOverlayTransformOffset()` there would be a bug. Neither site reads a clone's document position: `manualRowMove#refreshPositions` builds `tdStartPixel` from `THEAD.offsetHeight` plus `getRowsHeight(...)`, and `manualColumnMove#onMouseDown` builds `inlinePos` from `getColumnsWidth(...)`. Both are summed logical sizes in the master hider's space, so there is no layout-carried sticky shift to double-count — the transform variant is only for readers that walk the offset chain, which is why the editor is its only consumer.

`walkontable/AGENTS.md` does name both ("the `manualColumnMove` backlight over a frozen column, `manualRowMove` over a frozen row"). You are right that the PR description names only one; that is the author's to correct.

**3. Customer CSS on `.ht_master ~` — agreed, and not yet documented.**

A stylesheet reaching a clone through a sibling combinator stops matching once that clone sits in a rail, and nothing warns. Our own seam rule needed the parallel selector for exactly this reason. Flagging it for a changelog or migration note rather than deciding it here.

**4. `touchScroll` — the cached reference survives, and nothing is reparented mid-scroll.**

The cached node is `holder.parentNode`, which *is* the clone root — the same element `OverlayRail` holds — and the rail only ever moves it (`clone.before(rail)`, `rail.appendChild(clone)`, `rail.replaceWith(clone)`). It is never replaced or recreated, so the reference stays valid and `classList` travels with the node.

And `pin()` reparents only when it first builds the rail; on an already-pinned clone it rewrites styles and moves nothing. Pin state changes when an overlay starts or stops rendering, or an axis owner flips — neither happens during a momentum scroll on a window-scrolled grid.

**5. Manual QA — the author's call.** I do not touch the checkbox or the labels.

**6. `Overlay#destroy()` — I added the release, then reverted it.**

You are right that it is asymmetric with `reset()`. The problem is that it cannot be tested: I wrote the assertion (no `.htOverlayRail` in the document after `hot.destroy()`) and it passed *without* the fix, because `Core#destroy()` drops the whole root container and takes the rails with it. Rather than ship an untestable change plus a test that cannot fail, I left it and noted the trap. Happy to add it if you would rather have the symmetry.

**7. Forced-layout ordering — you are right, and this also answers the `pin()` write cost on the inline thread.**

I profiled both against a `develop` build: same fixture, same 40 wheel steps of 60px (2400px total), 5 runs each, medians from Chrome's own counters (CDP Performance domain). The numbers were stable to the count across runs.

Grid with frozen panes:

| | develop | this PR | delta |
|---|---|---|---|
| RecalcStyleCount | 209 | 289 | **+80** (+2 per draw) |
| RecalcStyleDuration | 7.4 ms | 8.4 ms | +1.0 ms |
| LayoutCount | 158 | 238 | **+80** (+2 per draw) |
| LayoutDuration | 52.0 ms | 53.4 ms | +1.4 ms |
| ScriptDuration | 116.7 ms | 121.3 ms | +4.6 ms |
| TaskDuration | 292.5 ms | 295.8 ms | +3.3 ms (**+1.1%**) |

Same scroll, column headers only, no frozen panes:

| | develop | this PR | delta |
|---|---|---|---|
| RecalcStyleCount | 169 | 169 | **0** |
| LayoutCount | 119 | 119 | **0** |
| LayoutDuration | 51.9 ms | 51.6 ms | 0 |
| TaskDuration | 271.2 ms | 267.7 ms | 0 |

Two things follow.

**`pin()` rewriting ~12 properties is not the cost.** The headers-only grid rails its top clone and calls `pin()` on every full draw, and the two builds are identical to the count — no extra style recalculation, no extra layout. So the early-out you suggested would buy nothing measurable, and I would rather not add cached placement state for it.

**The extra work is real and is where you predicted, on grids with frozen panes: exactly two more style recalculations and two more layouts per draw.** That is the forced-reflow signature — a layout-affecting write followed by a geometry read in the same draw, where the old transform write invalidated no layout. I have not isolated which read forces them, so I am not going to claim a specific line.

Size: +1.4 ms of layout across a 2400px, 40-frame scroll — about 0.035 ms per frame — and +1.1% of the scroll's total main-thread time. Real, and small enough that I would rather batch the reads properly as a follow-up than bolt a fix onto this PR. Say the word if you would rather it block here.

**Review by Haxikowy — APPROVED** [review]
